discovery/dockerswarm/nodes.go

```diff
@@ -0,0 +1,108 @@
+	swarmLabelNodeAvailability         = swarmLabelNodePrefix + "availability"
+	swarmLabelNodeEngineVersion        = swarmLabelNodePrefix + "engine_version"
+	swarmLabelNodeHostname             = swarmLabelNodePrefix + "hostname"
+	swarmLabelNodeID                   = swarmLabelNodePrefix + "id"
+	swarmLabelNodeLabelPrefix          = swarmLabelNodePrefix + "label_"
+	swarmLabelNodeManagerAddr          = swarmLabelNodePrefix + "manager_address"
+	swarmLabelNodeManagerLeader        = swarmLabelNodePrefix + "manager_leader"
+	swarmLabelNodeManagerReachability  = swarmLabelNodePrefix + "manager_reachability"
+	swarmLabelNodePlatformArchitecture = swarmLabelNodePrefix + "platform_architecture"
+	swarmLabelNodePlatformOS           = swarmLabelNodePrefix + "platform_os"
+	swarmLabelNodeRole                 = swarmLabelNodePrefix + "role"
+	swarmLabelNodeStatus               = swarmLabelNodePrefix + "status"
+)
+
+func (d *Discovery) refreshNodes(ctx context.Context) ([]*targetgroup.Group, error) {
+	tg := &targetgroup.Group{
+		Source: "DockerSwarm",
+	}
+
+	nodes, err := d.client.NodeList(ctx, types.NodeListOptions{Filters: makeArgs(d.filters)})
+	if err != nil {
+		return nil, fmt.Errorf("error while listing swarm nodes: %w", err)
+	}
+
+	for _, n := range nodes {
+		labels := model.LabelSet{
+			swarmLabelNodeID:                   model.LabelValue(n.ID),
+			swarmLabelNodeRole:                 model.LabelValue(n.Spec.Role),
+			swarmLabelNodeAvailability:         model.LabelValue(n.Spec.Availability),
+			swarmLabelNodeHostname:             model.LabelValue(n.Description.Hostname),
+			swarmLabelNodePlatformArchitecture: model.LabelValue(n.Description.Platform.Architecture),
+			swarmLabelNodePlatformOS:           model.LabelValue(n.Description.Platform.OS),
+			swarmLabelNodeEngineVersion:        model.LabelValue(n.Description.Engine.EngineVersion),
+			swarmLabelNodeStatus:               model.LabelValue(n.Status.State),
+			swarmLabelNodeAddress:              model.LabelValue(n.Status.Addr),
+		}
+		if n.ManagerStatus != nil {
+			labels[model.LabelName(swarmLabelNodeManagerLeader)] = model.LabelValue(fmt.Sprintf("%v", n.ManagerStatus.Leader))
+			labels[model.LabelName(swarmLabelNodeManagerReachability)] = model.LabelValue(n.ManagerStatus.Reachability)
+			labels[model.LabelName(swarmLabelNodeManagerAddr)] = model.LabelValue(n.ManagerStatus.Addr)
+		}
+
+		for k, v := range n.Spec.Labels {
+			ln := strutil.SanitizeLabelName(k)
+			labels[model.LabelName(swarmLabelNodeLabelPrefix+ln)] = model.LabelValue(v)
+		}
+
+		addr := net.JoinHostPort(n.Status.Addr, strconv.FormatUint(uint64(d.port), 10))
+		labels[model.AddressLabel] = model.LabelValue(addr)
+
+		tg.Targets = append(tg.Targets, labels)
+
+	}
+	return []*targetgroup.Group{tg}, nil
+}
+
+func (d *Discovery) getNodeLabels(ctx context.Context, nodeID string) (model.LabelSet, error) {
+	n, _, err := d.client.NodeInspectWithRaw(ctx, nodeID)
```

Comment:
This is a lot of RPCs, can we not get this from NodeList?

Similarly with the others.

Reply:
done